docs/proposals/push-based-metrics-collection.md

```diff
@@ -0,0 +1,196 @@
+    # It specifies the config of metrics collector, for example, 
+    # metrics_collector_config={"kind": "Push"},
+    metrics_collector_config: Dict[str, Any] = None, 
+)
+```
+
+### New Interface `report_metrics` in Python SDK
+
+```Python
+"""Push Metrics Directly to Katib DB
+    Katib DB Manager service should be accessible while calling this API.
+
+    If you run this API in-cluster (e.g. from the Kubeflow Notebook) you can
+    use the default Katib DB Manager address: `katib-db-manager.kubeflow:6789`.
+
+    If you run this API outside the cluster, you have to port-forward the
+    Katib DB Manager before getting the Trial metrics: `kubectl port-forward svc/katib-db-manager -n kubeflow 6789`.
+    In that case, you can use this Katib DB Manager address: `localhost:6789`.
+
+    You can use `curl` to verify that Katib DB Manager is reachable: `curl <db-manager-address>`.
+
+    [!!!] Trial name should always be passed into Katib Trials as env variable `KATIB_TRIAL_NAME`.
+
+    Args:
+        metrics: Dict of metrics pushed to Katib DB.
+            For examle, `metrics = {"loss": 0.01, "accuracy": 0.99}`.
+        db-manager-address: Address for the Katib DB Manager in this format: `ip-address:port`.
+    
+    Raises:
+        RuntimeError: Unable to push Trial metrics to Katib DB.
+"""
+def report_metrics(
+    metrics: Dict[str, Any],
+    db_manager_address: str = constants.DEFAULT_DB_MANAGER_ADDRESS,
+)
+```
+
+### A Simple Example:
+
+```Python
+import kubeflow.katib as katib
+
+# Step 1. Create an objective function with push-based metrics collection.
+def objective(parameters):
+    # Import required packages.
+    import kubeflow.katib as katib
+    # Calculate objective function.
+    result = 4 * int(parameters["a"]) - float(parameters["b"]) ** 2
+    # Push metrics to Katib DB.
+    katib.report_metrics({"result": result})
+
+# Step 2. Create HyperParameter search space.
+parameters = {
+    "a": katib.search.int(min=10, max=20),
+    "b": katib.search.double(min=0.1, max=0.2)
+}
+
+# Step 3. Create Katib Experiment with 12 Trials and 2 GPUs per Trial.
+katib_client = katib.KatibClient(namespace="kubeflow")
+name = "tune-experiment"
+katib_client.tune(
+    name=name,
+    objective=objective,
+    parameters=parameters,
+    objective_metric_name="result",
+    max_trial_count=12,
+    resources_per_trial={"gpu": "2"},
+    metrics_collector_config={"kind": "Push"},
+)
+
+# Step 4. Get the best HyperParameters.
+print(katib_client.get_optimal_hyperparameters(name))
+```
+
+## Implementation
+
+### Add New Parameter in `tune`
+
+As is mentioned above, we decided to add `metrics_collector_config` to the tune function in Python SDK. Also, we have some changes to be made:
+
+1. Disable injection: set `katib.kubeflow.org/metrics-collector-injection` to `disabled` when the push-based way of metrics collection is adopted so as to disable the injection of the metrics collection sidecar container.
```

Comment:
Do we really need it and where we are planning to disable it ?
Should we just verify if metricsCollector type is Push here: https://github.com/kubeflow/katib/blob/master/pkg/webhook/v1beta1/pod/inject_webhook.go#L102 ?

Reply:
Sorry for that, I think we do not really need it because there already exists some logic to handle this.

https://github.com/kubeflow/katib/blob/e6bd3e7b5bd95e0e77ce2f869e6b1eeb72d24180/pkg/webhook/v1beta1/pod/inject_webhook.go#L151

But we should not verify metricsCollector type in:

https://github.com/kubeflow/katib/blob/e6bd3e7b5bd95e0e77ce2f869e6b1eeb72d24180/pkg/webhook/v1beta1/pod/inject_webhook.go#L102

Because we need to add Katib Trial labels to the Pod metadata:

https://github.com/kubeflow/katib/blob/e6bd3e7b5bd95e0e77ce2f869e6b1eeb72d24180/pkg/webhook/v1beta1/pod/inject_webhook.go#L141

Reply:
That's right, we should mutate `KATIB_TRIAL_NAME` to the training container env.